crates/matrix-sdk/src/client/login_builder.rs

```diff
@@ -0,0 +1,313 @@
+
+    fn tracing_desc(&self) -> &'static str {
+        match self {
+            LoginMethod::UserPassword { .. } => "identifier and password",
+            LoginMethod::Token(_) => "token",
+        }
+    }
+
+    fn to_login_info(&self) -> login::v3::LoginInfo<'a> {
+        match self {
+            LoginMethod::UserPassword { id, password } => {
+                login::v3::LoginInfo::Password(login::v3::Password::new(id.clone(), password))
+            }
+            LoginMethod::Token(token) => login::v3::LoginInfo::Token(login::v3::Token::new(token)),
+        }
+    }
+}
+
+/// Builder type used to configure optional settings for logging in with a
+/// username or token.
+///
+/// Created with [`Client::login_username`] or [`Client::login_token`].
+/// Finalized with [`.exec()`](Self::exec).
+#[allow(missing_debug_implementations)]
+pub struct LoginBuilder<'a> {
+    client: Client,
+    login_method: LoginMethod<'a>,
+    device_id: Option<&'a str>,
+    initial_device_display_name: Option<&'a str>,
+}
+
+impl<'a> LoginBuilder<'a> {
+    fn new(client: Client, login_method: LoginMethod<'a>) -> Self {
+        Self { client, login_method, device_id: None, initial_device_display_name: None }
+    }
+
+    pub(super) fn new_password(client: Client, id: UserIdentifier<'a>, password: &'a str) -> Self {
+        Self::new(client, LoginMethod::UserPassword { id, password })
+    }
+
+    pub(super) fn new_token(client: Client, token: &'a str) -> Self {
+        Self::new(client, LoginMethod::Token(token))
+    }
+
+    /// Set the device ID.
+    ///
+    /// The device ID is a unique id that will be associated with this session.
+    /// If not set, the homeserver will create one. Can be an existing device ID
+    /// from a previous login call. Note that this should be done only if the
+    /// client also holds the corresponding encryption keys.
+    pub fn device_id(mut self, value: &'a str) -> Self {
+        self.device_id = Some(value);
+        self
+    }
+
+    /// Set the initial device display name.
+    ///
+    /// The device display name is the public name that will be associated with
+    /// the device_id. Only necessary the first time you login with this device
+    /// ID. It can be changed later.
+    pub fn initial_device_display_name(mut self, value: &'a str) -> Self {
+        self.initial_device_display_name = Some(value);
+        self
+    }
+
+    /// Send the login request.
+    #[instrument(
+        target = "matrix_sdk::client",
+        name = "login",
+        skip_all,
+        fields(method = self.login_method.tracing_desc()),
+    )]
+    pub async fn exec(self) -> Result<login::v3::Response> {
+        let homeserver = self.client.homeserver().await;
+        info!(homeserver = homeserver.as_str(), identifier = ?self.login_method.id(), "Logging in");
+
+        let request = assign!(login::v3::Request::new(self.login_method.to_login_info()), {
+            device_id: self.device_id.map(|d| d.into()),
```

Comment:
Could be simplify as:

```suggestion
            device_id: self.device_id.map(Into::into),
```

It saves one closure.